crates/flipperzero/src/xbm.rs

```diff
@@ -0,0 +1,479 @@
+
+        assert!(bytes as usize == SIZE, "dimensions don't match data length");
+
+        Self {
+            data: ByteArray(data),
+            width: WIDTH,
+            height: HEIGHT,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct ByteArray<const N: usize>(pub [u8; N]);
+
+impl<const N: usize> Deref for ByteArray<N> {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        self.0.as_slice()
+    }
+}
+
+impl<const N: usize> DerefMut for ByteArray<N> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0.as_mut_slice()
+    }
+}
+
+/// Creates a compile-time XBM image.
+/// The type of this expression is [`XbmImage`] with [`ByteArray`] backend of the calculated size.
+///
+/// The syntax is an [XBM image definition][XBM format]
+/// optionally wrapped in an `unsafe` block.
+///
+/// Unless the expression is wrapped in an `unsafe` block
+/// the macro will perform identifier validation:
+/// it will check that identifiers are in the order
+/// `<name>_width`, `<name>_height`, optionally `<name>_x_hot` and `<name>_y_hot` and `<name>_bits`
+/// and that the `<name>` is the same for all identifiers.
+/// The `unsafe` form omits this validations
+/// while still ensuring that the size constraints are valid.
+///
+/// [XBM format]: https://www.fileformat.info/format/xbm/egff.htm
+#[macro_export]
+macro_rules! xbm {
+    (
+        #define $width_ident:ident $width:literal
+        #define $height_ident:ident $height:literal
+        $(
+            #define $x_hotspot_ident:ident $_hotspot_x:literal
+            #define $y_hotspot_ident:ident $_hotspot_y:literal
+        )?
+        static $(unsigned)? char $bits_ident:ident[] = {
+            $($byte:literal),* $(,)?
+        };
+    ) => {{
+        { // name assertions
+            let bits_ident = stringify!($bits_ident).as_bytes();
+            assert!(
+                matches!(bits_ident, [.., b'_', b'b', b'i', b't', b's']),
+                "width identifier should end with `_bits`",
+            );
+            let significant_len = bits_ident.len() - b"_bits".len();
+
+            const fn str_eq(left: &[u8], right: &[u8], limit: usize) -> bool {
+                match (left.split_first(), right.split_first()) {
+                    (
+                        Some((&left_first, left_remaining)),
+                        Some((&right_first, right_remaining)),
+                    ) => {
+                        left_first == right_first
+                            && (limit == 1 || str_eq(left_remaining, right_remaining, limit - 1))
+                    }
+                    (None, None) => true,
+                    _ => false,
+                }
+            }
+
+            let width_ident = stringify!($width_ident).as_bytes();
+            assert!(
+                matches!(width_ident, [.., b'_', b'w', b'i', b'd', b't', b'h']),
+                "width identifier should end with `_width`",
+            );
```

Comment:
I wonder if we should replace these with [`static_assertions`](https://crates.io/crates/static_assertions) (not necessarily in this PR).

I think the asserts make sense, though I do worry about the runtime/code-size overhead that we might be taking on. Probably worth doing a side-by-side with/without those asserts and see how much additional code they introduce.

Unfortunately binary size has a very direct impact on how quickly Flipper applications start.

Reply:
Generally, this would normally be replaced with a proc-macro in the future.

As for now, I think that it should normally be remomved by the compiler, considering that it is intended for use in `const`s where the expression has to be CT-evaluated.